### ISSUE TITLE
Errors thrown for `no-inferred-empty-object-type` & `strict-boolean-expressions` rules

### ISSUE THREAD

**Issue by chrismbarr** [palantir/tslint#1909]
### Bug Report

- __TSLint version__: `4.1.1`
- __TypeScript version__: `2.1.4`
- __Running TSLint via__: `gulp-tslint`

#### with `tslint.json` configuration:

```json
"no-inferred-empty-object-type": true,
"strict-boolean-expressions": true,
```

#### with `tsconfig.json` configuration: 
```json
{
    "compileOnSave": true,
    "compilerOptions": {
        "allowUnreachableCode": false,
        "allowUnusedLabels": false,
        "forceConsistentCasingInFileNames": true,
        "module": "commonjs",
        "noFallthroughCasesInSwitch": true,
        "noImplicitAny": true,
        "noImplicitReturns": true,
        "noImplicitThis": true,
        "noImplicitUseStrict": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "preserveConstEnums": true,
        "removeComments": true,
        "sourceMap": false,
        "strictNullChecks": true,
        "target": "es5"
    },
    "include": [
        "src",
        "tests"
    ]
}
```

#### `gulpfile.js` task _(with typechecking)_
```js
const gulp        = require("gulp");
const gulpTslint  = require("gulp-tslint");
const tslint      = require("tslint");

gulp.task("lint", () => {
    const tslintProgram = tslint.Linter.createProgram("./tsconfig.json", ".");

    const tsResult = tsProject.src()
        .pipe(gulpTslint({
            formattersDirectory: "src",
            formatter: "stylish",
            program: tslintProgram
        }))
        .pipe(gulpTslint.report({
            emitError: false,
            reportLimit: 50
        }))
        .pipe(tsProject());

    return tsResult.js.pipe(gulp.dest("./"));
});
```

#### Actual behavior
This error is thrown
```
[07:04:46] Starting 'lint'...
stream.js:74
      throw er; // Unhandled stream error in pipe.
      ^

TypeError: Cannot read property 'name' of undefined
    at getSignatureFromDeclaration (C:\<project>\node_modules\typescript\lib\typescript.js:28685:47)
    at getThisTypeOfDeclaration (C:\<project>\node_modules\typescript\lib\typescript.js:27296:43)
    at checkThisExpression (C:\<project>\node_modules\typescript\lib\typescript.js:33759:32)
    at checkExpressionWorker (C:\<project>\node_modules\typescript\lib\typescript.js:37916:28)
    at checkExpression (C:\<project>\node_modules\typescript\lib\typescript.js:37894:42)
    at checkNonNullExpression (C:\<project>\node_modules\typescript\lib\typescript.js:35270:24)
    at checkPropertyAccessExpressionOrQualifiedName (C:\<project>\node_modules\typescript\lib\typescript.js:35317:24)
    at checkPropertyAccessExpression (C:\<project>\node_modules\typescript\lib\typescript.js:35284:20)
    at checkExpressionWorker (C:\<project>\node_modules\typescript\lib\typescript.js:37937:28)
    at checkExpression (C:\<project>\node_modules\typescript\lib\typescript.js:37894:42)
```

----

Just to be clear, the error **is not thrown** when I configure the rules like this:

```json
"no-inferred-empty-object-type": false,
"strict-boolean-expressions": false,
```

When **either one** (or both) of those rules are set to be `true` I get the above error.

Every time I try to narrow this down to only build/lint against fewer files with `tsconfig.json` by setting the `files` or `exclude` property like this:

```json
    "include": [
        "src",
        "tests"
    ],
    "exclude": [
        "tests"
    ]
```

or even 
```json
    "include": [
        "src"
    ]
```

It throws this error: 
```
[08:18:07] Starting 'lint'...
stream.js:74
      throw er; // Unhandled stream error in pipe.
      ^

Error:
Invalid source file: AlphabeticallyFirstFile.ts. Ensure that the files supplied to lint have a .ts, .tsx, .js or .jsx extension.

    at Linter.getSourceFile (C:\<project>\node_modules\tslint\lib\linter.js:186:19)
    at Linter.getEnabledRules (C:\<project>\node_modules\tslint\lib\linter.js:156:31)
    at Linter.lint (C:\<project>\node_modules\tslint\lib\linter.js:83:33)
    at C:\<project>\node_modules\gulp-tslint\index.js:102:16
    at wrappedMapper (C:\<project>\node_modules\map-stream\index.js:84:19)
    at Stream.stream.write (C:\<project>\node_modules\map-stream\index.js:96:21)
    at DestroyableTransform.ondata (C:\<project>\node_modules\through2\node_modules\readable-stream\lib\_stream_readable.js:531:20)
    at emitOne (events.js:96:13)
    at DestroyableTransform.emit (events.js:188:7)
    at readableAddChunk (C:\<project>\node_modules\through2\node_modules\readable-stream\lib\_stream_readable.js:198:18)
```

_(Note that `AlphabeticallyFirstFile.ts` is an example name here... but it always reports on the first file listed in the folder I include here.)_  This **does not happen** if i include both of the folders like in my original example above.

--- 

Overall... I'm not sure why these rules throw errors, and I'm not sure why I can't narrow down my project code to only lint certain files to give a code example.  Any help here would be greatly appreciated, but for now I'm going to have to disable these rules until they work again.
